src/CodeGeneration.Roslyn.Engine/CompilationGenerator.cs

```diff
@@ -34,29 +33,48 @@ public class CompilationGenerator
         private readonly Dictionary<string, (PluginLoader loader, Assembly assembly)> cachedPlugins = new Dictionary<string, (PluginLoader, Assembly)>(StringComparer.OrdinalIgnoreCase);
 
         /// <summary>
-        /// Gets or sets the list of paths of files to be compiled.
+        /// Initializes a new instance of the <see cref="CompilationGenerator"/> class.
         /// </summary>
-        public IReadOnlyList<string> Compile { get; set; }
+        /// <param name="projectDirectory">The list of paths of files to be compiled.</param>
+        /// <param name="compile">The directory with the project file.</param>
+        /// <param name="referencePath">The list of paths to reference assemblies.</param>
+        /// <param name="preprocessorSymbols">A set of preprocessor symbols to define.</param>
+        /// <param name="pluginPaths">The paths to plugins.</param>
+        /// <param name="intermediateOutputDirectory">The path to the directory that contains generated source files.</param>
+        public CompilationGenerator(string projectDirectory, IReadOnlyList<string> compile, IReadOnlyList<string> referencePath, IEnumerable<string> preprocessorSymbols, IReadOnlyList<string> pluginPaths, string intermediateOutputDirectory)
+        {
```

Comment:
This is technically a public entrypoint so I would add input validation to these arguments. 
However if the use case is that this is only public for calling between assemblies defined in this repo (i.e. the assembly isn't a ref assembly that we offer in a nuget package), and the caller also has NRT turned on, then we can omit them here.